### ISSUE TITLE
In what case does syslog-ng cannot parse source message

### ISSUE THREAD

**Issue by aniketpant1** [syslog-ng/syslog-ng#3755]
I have seen this case where syslog-ng cannot parse source message like if device is sending logs to syslog-ng then syslog-ng parse only half of the message i think the reason might be dot `.` present in the message.

**Comment by MrAnno** [comment]
I believe this is the same question that has been answered in #3749.

https://github.com/syslog-ng/syslog-ng/issues/3749#issuecomment-897658919

**Comment by MrAnno** [comment]
Check whether the “dot” you’re seeing  is a NULL character. If so, I would rather try to fix this on the other side where nulls are produced, because those NULLs are going to cause problems everywhere.
Switching from `network()` TCP source to `syslog()` and disabling all message processing and parsing might produce a similar state that you had with UDP sources, but then you will probably start having problems on the destination side.

In summary, if you have nulls in your messages, try fixing/removing them, because they will haunt you everywhere you forward them.

**Comment by aniketpant1** [comment]
Yes but i didn't get the proper solution that time so i have to raise the another question but now we can close this issue because i got the issue why it was all happening

**Comment by aniketpant1** [comment]
But i am using transport(proxied-tcp) protocol

**Comment by aniketpant1** [comment]
Sorry things i forget to mention that in UDP case this string termination
by NUL character does not happen

**Comment by MrAnno** [comment]
Yep, you can achieve the same with the `syslog()` source (with `flags(no-parse)`), which implements RFC6587 and uses frames instead of delimiters to separate messages. But as I mentioned on Stack Overflow and here as well, those nulls will probably bother you everywhere else, so the best would be if you could remove them. You can find an example rewrite rule in the Stack Overflow answer that removes nulls before further processing.

**Comment by aniketpant1** [comment]
@MrAnno 
```
source s_network_tcp {
    syslog (
	port(1470)
	transport("tcp")
	flags(no-parse)
	);
};

destination ise_kafka_tcp
{
file(
            "/var/log/ise/${YEAR}.${MONTH}.${DAY}logs.log"
            owner("root")
            group("root")
            perm(0777)
    );
};
log{ source(s_network_tcp); destination(ise_kafka_tcp);};
```
Output of syslog-ng -Fedtv
```
[2021-08-16T11:52:25.552957] syslog-ng starting up; version='3.31.2'
[2021-08-16T11:52:25.552963] Running application hooks; hook='2'
[2021-08-16T11:52:41.820197] Syslog connection accepted; fd='16', client='AF_INET(xx.xx.xx.xx:49756)', local='AF_INET(0.0.0.0:1470)'
[2021-08-16T11:52:45.870850] Invalid frame header; header=''
[2021-08-16T11:52:45.870925] Syslog connection closed; fd='16', client='AF_INET(xx.xx.xx.xx:49756)', local='AF_INET(0.0.0.0:1470)'
[2021-08-16T11:52:45.870977] Closing log transport fd; fd='16'
[2021-08-16T11:53:15.874756] Syslog connection accepted; fd='17', client='AF_INET(xx.xx.xx.xx:49860)', local='AF_INET(0.0.0.0:1470)'
[2021-08-16T11:53:17.837703] Invalid frame header; header=''
[2021-08-16T11:53:17.837787] Syslog connection closed; fd='17', client='AF_INET(xx.xx.xx.xx:49860)', local='AF_INET(0.0.0.0:1470)'
[2021-08-16T11:53:17.837848] Closing log transport fd; fd='17'
```
no messages  appearing now and yes if i use `transport("proxied-tcp") in syslog source then message will stream

**Comment by aniketpant1** [comment]
@MrAnno this process in not working 
for removing null character i made a file but syslog-ng does not processing regex pattern
this is testing file [NULTEST.log](https://github.com/syslog-ng/syslog-ng/files/7024736/NULTEST.log)

```
source s_problematic_message {
  file("/etc/syslog-ng/NULTEST.log"
  flags(no-parse)
  );
};
rewrite r_remove_nulls {
  subst('\x00',"", value("MESSAGE"), type(pcre), flags(global)); # single quotes!
};

destination d_output {
 file("/var/log/ise/ise.log");
};

log {
  source(s_problematic_message);
  rewrite(r_remove_nulls);
  destination(d_output);
};

```
I used single quote in replacement field but the problem is same 
can you please tell me why it is not working

**Comment by MrAnno** [comment]
`r_remove_nulls` works only with sources that do NOT separate messages based on line endings (`\n`, `\r`, `NULL`).
For example, it works with `udp()`, and `syslog()`:

```
log {
    source { udp(port(4444)); };
    
    rewrite {
        subst('\x00',"", value("MESSAGE"), type(pcre), flags(global));
    };

    destination {
        file("/dev/stdout");
    };
};
```

```
$ nc -u 127.0.0.1 4444 < ~/Downloads/NULTEST.log

[2021-08-21T07:49:24.085279] Incoming log entry; line='Name=Aniket Pant,Ip=192.168.0.114,Id='
[2021-08-21T07:49:24.085589] Outgoing message; message='Aug 21 07:49:24 localhost Name=Aniket Pant,Ip=192.168.0.114,Id=E=541\x0a'
```

**Comment by aniketpant1** [comment]
so will it work for
```
 syslog(
    transport("tcp")
    port(4444)
	flags(no-parse)
);
```

**Comment by MrAnno** [comment]
Yes, the example is above.

**Comment by bazsi** [comment]
As far as I remember the NUL character in tcp would split the message at
that character, just as an NL would.

**Comment by MrAnno** [comment]
Without `syslog()`'s framing, it is currently not possible to keep a message intact that has a NULL character inside of it, because as Bazsi mentioned it, NUL would split the message.

**Comment by MrAnno** [comment]
If you switch to `syslog()`, the sender has to use the RFC6587 transport format too.

TLDR:
- The `network()` source treats NULL characters as if they were a message boundaries.
- The `syslog()` source handles NULL characters transparently with the `flags(no-parse)` flag, but it requires RFC6587 formatting on the sender's side.

**Comment by aniketpant1** [comment]
Hey @MrAnno i have discuss with CISCO engineer there were saying that CISCO ISE is sending complete message we can see the complete message in wireshark.It is difficult to change the rfc for one particular use case the dot character one.They use rfc 3164 for syslog is there any other alternative tool we can process dot character or without any changing in rfc we can process the message.

**Comment by MrAnno** [comment]
Just to make it clear, we are talking about NULL (`\0`) characters, not dots.

>  is there any other alternative tool we can process dot character

If you can remove those null bytes from the log messages before reaching syslog-ng (by any external tools, or custom script), it will work as expected.

Currently, NULLs are handled as message boundaries in the case of rfc3164 network sources, but we could open an "enhancement" request about an option that would either remove or just allow those null-bytes inside a message.

Are those valid messages? I mean embedding nulls inside logs seems unintentional, it sounds like a bug on the producer side.

**Comment by aniketpant1** [comment]
This NULL is streaming from ISE.They are not gonna remove this NUL it is happening from the some programming language. If they change the rfc they have to check compliance of the rfc6587 for both udp and tcp.It is a long process.

Do you know any external tools or script we can remove this NUL before it reach to syslog.

>  but we could open an "enhancement" request about an option that would either remove or just allow those null-bytes inside a message.
If this can happen it would be great.

**Comment by aniketpant1** [comment]
Hey @bazsi i will try this

**Comment by MrAnno** [comment]
#3913 has been merged.

The `transport(text-with-nuls)` option will be part of the next syslog-ng release.
There are deb/rpm packages available for testing purposes if you want to try this new feature quickly.

**Comment by aniketpant1** [comment]
> There are deb/rpm packages available for testing purposes if you want to try this new feature quickly.

What version i have to install ?

**Comment by aniketpant1** [comment]
I've downloaded wrong syslog-ng version 3.35, i need to download version 3.36 where i can get this version.I downloaded from [https://www.syslog-ng.com/products/open-source-log-management/3rd-party-binaries.aspx](https://www.syslog-ng.com/products/open-source-log-management/3rd-party-binaries.aspx) then i go to EPEL --> [Weekly build from latest (unreleased) master](https://www.syslog-ng.com/community/b/blog/posts/rpm-packages-from-syslog-ng-git-head)

I got this issue
```
WARNING: With use-dns(no), dns-cache() will be forced to 'no' too!;
72552] Unknown value specified in the transport() option, no such LogProto plugin found; transport='text-with-nuls'
72561] Error initializing message pipeline; plugin_name='network', location='/etc/syslog-ng/conf.d/ise-test.conf:2:4'
```

**Comment by czanik** [comment]
When did you try it? I finished building the git snapshot packages just 20 minutes ago...

**Comment by aniketpant1** [comment]
> When did you try it? I finished building the git snapshot packages just 20 minutes ago...

15-20 mintues ago i think

**Comment by MrAnno** [comment]
v3.36.1 has not been released yet.
You can try @czanik's snapshot repo, or download binaries from our CI:
https://github.com/syslog-ng/syslog-ng/actions/runs/1852785401
(or, as the last alternative, build from source)

**Comment by czanik** [comment]
@aniketpant1: my latest build is 3.35.1.353.gcdd0351. It should have everything included up until yesterday in the evening.

**Comment by aniketpant1** [comment]
![Screenshot 2022-02-17 153516](https://user-images.githubusercontent.com/40541079/154453149-dfb7c40e-f5c2-460f-9149-9ee7ea2c6ca4.jpg)
do i have to install all the rpm package in my linux machine or just `syslog-ng-3.35.1.353.gcdd0351-2+20220216T123454.src.rpm`

**Comment by aniketpant1** [comment]
hey @czanik it is now downloaded version 3.35.1.353.Now i will test the dot character logs.

**Comment by aniketpant1** [comment]
When this version release.

**Comment by bazsi** [comment]
On Fri, Feb 18, 2022, 07:43 aniketpant1 ***@***.***> wrote:

> Hey @czanik <https://github.com/czanik> , @MrAnno
> <https://github.com/MrAnno> , @bazsi <https://github.com/bazsi>
> I have tested the logs under this following cases
> **Case 1 **
> no proxied-tcp in source
>
> source s_network_tcp {
>    network(port(1470) transport(text-with-nuls) flags(no-multi-line));
> };
> destination ise_kafka
> {
> file(
>   "/var/log/ise/ise.log"
> };
>
> *Result of Case 1*
> NUL character is removed ; the message is not truncated as it was
> happening earlier.
>
> Feb 18 10:28:45 test-host CISE_Failed_Attempts 0001830934 4 3  Response={RadiusPacketType=AccessReject; AuthenticationResult=Failed; MS-CHAP-Error= E=691 R=0 C=4BEEC70B01B76E40C7602F1CAAC2B759 V=3; },
>
> **Case 2 **
> If i used proxied-tcp protocol in source.
>
> source s_network_tcp {
>    network(port(1470) transport(text-with-nuls) transport("proxied-tcp") flags(no-multi-line));
> };
>
> filter src4{host("test-host")  };
> rewrite ise_tag {
> set("ise",value("tags"));
> set("$PROXIED_SRCIP",value("HOST_FROM"));
> };
>
> destination ise_kafka
> {
> file(
>   "/var/log/ise/ise.log"
>   template("$(format-json time=$ISODATE tags proxied_srcip=$PROXIED_SRCIP  host=$HOST message='${MSGHDR}${MSG}')\n")
>
>   );
> };
> log{ source(s_network_tcp); destination(ise_kafka);};
>
> *Result of Case 2*
>
> {"time":"2022-02-18T12:06:22+05:30","proxied_srcip":"xx.xx.xx.xx","message":"CISE_Failed_Attempts 0007451812 4 3  IPSEC=IPSEC#Is IPSEC Device, EnableFlag=Enabled, Response={RadiusPacketType=AccessReject; AuthenticationResult=Failed; MS-CHAP-Error=","host":"test-host"}
> {"time":"2022-02-18T12:06:22+05:30","proxied_srcip":"xx.xx.xx.xx","message":"E=691 R=0 C=1D78FD6000FF2489FBFE2FD472BBE666 V=3; },","host":"different_host"}
>
> Message get truncated if i used proxied-tcp protocol.
>

You can't use two transport() options, it's one or the other. NUL support
was only added to as a separate transport() implementation.

My understanding was that you receive this directly from a router.

Bazsi

**Comment by aniketpant1** [comment]
Thanks @bazsi when i will see this syslog-ng version 3.35.1.353 stable.
